include/pipe.h

```diff
@@ -0,0 +1,93 @@
+#ifndef PIPE_NETWORK_PIPE_H_
+#define PIPE_NETWORK_PIPE_H_
+
+#include "node.h"
+#include <array>
+#include <math.h>
+#include <memory>
+
+//! Pipe class
+//! \brief Class that stores the information about pipes
+class Pipe {
+
+ public:
+  // Constructor with id and node pointers
+  //! \param[in] id node id
+  //! \param[in] nodes array of node pointers
+  Pipe(unsigned id, const std::array<std::shared_ptr<Node>, 2>& nodes)
+      : id_{id}, array_node_ptr_{nodes} {}
+
+  //! Destructor
+  ~Pipe() { array_node_ptr_.fill(nullptr); }
+
+  //! Copy constructor
+  Pipe(const Pipe&) = delete;
+
+  //! Assignment operator
+  Pipe& operator=(const Pipe&) = delete;
+
+  //! Move constructor
+  Pipe(Pipe&&) = delete;
+
+  //! Return id
+  //! \retval id_ id of the pipe
+  unsigned id() const { return id_; }
+
+  //! Return array of node pointers
+  //! \retval array_node_ptr_ array of node pointers
+  std::array<std::shared_ptr<Node>, 2> array_node_ptr() const {
```

Comment:
You shouldn't return nodes this is giving up control. So philosophically, this shouldn't happen. Pipe should own the node objects, you can't pass it to something else.